lib/marin/src/marin/rl/environments/inference_ctx/base.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2025 The Marin Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Inference context for rollout construction.
+
+This context is provided to environments and provides access to the inference server
+as well as methods for tokenization and logprob extraction from an OpenAI ChatCompletion.
+"""
+
+import logging
+import jax.numpy as jnp
+import numpy as np
+from openai.types.chat import ChatCompletion
+from openai.types.chat.chat_completion import Choice
+from marin.rl.types import Rollout
+from levanter.models.lm_model import LmHeadModel
+
+logger = logging.getLogger(__name__)
+
+
+class BaseInferenceContext:
+    """Base class for inference contexts."""
+
+    def reload_model(self, model: LmHeadModel) -> None:
+        raise NotImplementedError
+
+    def batch_completions(
+        self,
+        prompts: list[str],
+        temperature: float,
+        n: int,
+        max_tokens: int | None = None,
+        stop: list[str] | None = None,
+    ) -> list[ChatCompletion]:
+        """Batch completions from the inference server."""
+        raise NotImplementedError
+
+    def tokenize_prompt(self, prompt: str, choice: Choice | None = None) -> np.ndarray:
+        """Tokenize with chat template matching server behavior."""
+        messages = [{"role": "user", "content": prompt}]
+        try:
+            tokens = self.tokenizer.apply_chat_template(messages, tokenize=True, add_generation_prompt=True)
+        except Exception as e:
+            logger.warning(f"Chat template failed: {e}")
+            prompt_text = "\n".join([f"{msg['role']}: {msg['content']}" for msg in messages])
+            tokens = self.tokenizer.encode(prompt_text, add_special_tokens=True)
+            if not tokens:
+                raise ValueError(f"Failed to tokenize: {prompt[:100]}...") from None
+
+        return np.array(tokens, dtype=np.int32)
+
+    def tokenize_response(self, text: str) -> np.ndarray:
```

Comment:
It's a little weird/confusing to have tokenize_response which uses encode, and then response_tokens_from_choice which uses  convert_tokens_to_ids.

Which should we use, when? I'd maybe lean towards hiding these in favor of the higher-level APIs, since this can be subtle and easy to break.